src/libs/data-table-provenance.js

```diff
@@ -0,0 +1,90 @@
+import _ from 'lodash/fp'
+import { useCallback, useState } from 'react'
+import { Ajax } from 'src/libs/ajax'
+import { reportError } from 'src/libs/error'
+import { useCancellation } from 'src/libs/react-utils'
+import * as Utils from 'src/libs/utils'
+
+
+export const maxSubmissionsQueriedForProvenance = 25
+
+const getSubmissionsWithRootEntityType = async (workspace, entityType, { signal } = {}) => {
+  const { workspace: { namespace, name } } = workspace
+
+  // The root entity type is stored on method configurations, not submissions. To avoid fetching
+  // every submission's configuration, first use the list submissions endpoint and filter by
+  // submission entity type.
+  //
+  // If a workflow is run on a single entity, the submission entity type will be the root entity type.
+  // If a workflow is run on multiple entities, the submission entity type will be a set of the root entity type.
+  //
+  // After fetching method configurations, another filter is required remove submissions whose root entity type
+  // was a set of the desired entity type.
+  const allSubmissions = await Ajax(signal).Workspaces.workspace(namespace, name).listSubmissions()
+  const submissionsMaybeUsingEntityType = _.flow(
+    _.filter(submission => {
+      const submissionEntityType = submission.submissionEntity.entityType
+      return submissionEntityType === entityType || submissionEntityType === `${entityType}_set`
+    }),
+    _.slice(0, maxSubmissionsQueriedForProvenance)
+  )(allSubmissions)
+
+  const submissionConfigurations = await Promise.all(
+    _.map(
+      submission => Ajax(signal).Workspaces.workspace(namespace, name).submission(submission.submissionId).getConfiguration(),
+      submissionsMaybeUsingEntityType
+    )
+  )
+
+  return _.flow(
+    _.map(([submission, configuration]) => ({ submission, configuration })),
+    _.filter(({ configuration }) => configuration.rootEntityType === entityType),
+    _.orderBy(({ submission }) => new Date(submission.submissionDate), 'desc')
+  )(_.zip(submissionsMaybeUsingEntityType, submissionConfigurations))
+}
+
+const getColumnFromOutputExpression = expr => {
+  const match = /^this\.([^.]+)$/.exec(expr)
+  return match ? match[1] : null
+}
+
+const getColumnProvenance = async (workspace, entityType, { signal } = {}) => {
+  const submissions = await getSubmissionsWithRootEntityType(workspace, entityType, { signal })
+
+  return _.flow(
+    _.flatMap(({ submission, configuration }) => _.map(([output, expression]) => ({ submission, configuration, output, expression }), _.toPairs(configuration.outputs))),
+    _.map(({ submission, configuration, output, expression }) => ({ submission, configuration, output, column: getColumnFromOutputExpression(expression) })),
```

Comment:
Then here, you could handle the `Left` case in the context of where it is used.
```javascript
column: getColumnFromOutputExpression(expression).getOr(null)
```
I like this because it is here where I know best what to do with the failure case.

Now, as for the line, what I think is happening here is that you are essentially converting `expression` to `column`, but the line is a bit noisy for that to be obvious. I might write it this way to give the transformation more signal:
```javascript
_.map(x => ({ ..._.omit(['expression'], x), column: getColumnFromOutputExpression(expression) }))
```

Reply:
Changed the latter in 959268e1cfa7295c1d3848b5e4f86a37413fdb13. I think part of the reason I had the arguments destructured was because I was unsure what to name a single argument. In retrospect though, appropriate names seem fairly obvious.

Reply:
I hear that.

Not to belabor this, but I was reading the destructuring docs yesterday and it made me think you could do patterns like this without the `omit` (which struck me because it's a common pattern):
```javascript
_.map(({expression ...rest}) => ({ ...rest, column: getColumnFromOutputExpression(expression) }))
```

Reply:
I like it. Seems easier to read than `..._.omit`. Changed in https://github.com/DataBiosphere/terra-ui/pull/3332/commits/5ff7d5d67d3e5e323c636677799b025c74c2432c.